### PR TITLE
CMDCT-6350: Connect dropdown error messages to field

### PR DESCRIPTION
**Pull request by amanda-forrester**
### Description

Error messages for the dropdown field are not connected to the input, therefore Assistive Tech (AT) will not announce the error message. This PR programmatically connects error message to input field via aria-describedby and aria-invalid.

### Related ticket(s)

CMDCT-6350

---
### How to test

- Run this PR locally
- Navigate to NAAAR -> Add/copy NAAAR
- With the Add/copy modal open, [turn on VoiceOver](https://support.apple.com/guide/voiceover/turn-voiceover-on-or-off-vo2682/mac) (if you're on macOS)
- Use the Tab key to navigate to the required Program name field or another required dropdown field
- Once you blur one of the required fields, return to the errored field, you should be able to hear that the error message is read
- You should be able to hear the error message when it's visible and the field is focused

### Notes

See [this PR](https://github.com/Enterprise-CMCS/macpro-mdct-mfp/pull/1476) for similar work in MFP

---
### Pre-review checklist

- [ ] I have added [thorough](https://shorturl.at/aejkF) tests, if necessary
- [ ] I have updated relevant documentation, if necessary
- [ ] I have performed a self-review of my code
- [ ] I have manually tested this PR in the deployed cloud environment

---
#### Review
- [ ] Design: This work has been reviewed and approved by design, if necessary
- [ ] Product: This work has been reviewed and approved by product owner, if necessary

#### Security
_If either of the following are true, notify the team's ISSO (Information System Security Officer)._

- [ ] These changes are significant enough to require an update to the SIA.
- [ ] These changes are significant enough to require a penetration test.

### CODE REVIEW

File: services/ui-src/src/components/fields/DropdownField.test.tsx [hunk 1]
```diff
@@ -313,4 +313,34 @@ describe("<DropdownField />", () => {
...
+          },
+        },
+      }));
+
+      render(dropdownComponentWithOptions);
+      const dropdown = screen.getByLabelText("test-dropdown-label");
+
+      expect(dropdown).toHaveAttribute("aria-invalid", "true");
+
+      const describedById = dropdown.getAttribute("aria-describedby");
```
Comment (bangbay-bluetiger) [inline services/ui-src/src/components/fields/DropdownField.test.tsx]: This assertion can be written like the one above it:
```
expect(dropdown).toHaveAttribute("aria-describedby", "testDropdown__error");
```

File: services/ui-src/src/components/fields/DropdownField.test.tsx [hunk 2]
```diff
@@ -313,4 +313,34 @@ describe("<DropdownField />", () => {
...
+      expect(describedById).toContain("testDropdown__error");
+
+      const errorEl = document.getElementById("testDropdown__error");
```
Comment (bangbay-bluetiger) [inline services/ui-src/src/components/fields/DropdownField.test.tsx]: I would avoid `document.getElementById` and use React testing library methods. You could do something like this:
```
const errorEl = screen.getByText("A response is required");
expect(errorEl).toHaveAttribute("id", "testDropdown__error");
```

File: services/ui-src/src/components/fields/DropdownField.test.tsx [hunk 3]
```diff
@@ -313,4 +313,34 @@ describe("<DropdownField />", () => {
   testA11yAct(dropdownComponentWithOptions, () => {
     mockGetValues(undefined);
   });
+
+  describe("Dropdown connects error message to select for accessibility", () => {
+    afterEach(() => {
+      jest.clearAllMocks();
```
Comment (bangbay-bluetiger) [inline services/ui-src/src/components/fields/DropdownField.test.tsx]: There aren't any calls that need to be reset in this block, so you can remove this `afterEach`.

File: services/ui-src/src/components/fields/DropdownField.test.tsx [hunk 4]
```diff
@@ -313,4 +313,34 @@ describe("<DropdownField />", () => {
   testA11yAct(dropdownComponentWithOptions, () => {
     mockGetValues(undefined);
   });
+
+  describe("Dropdown connects error message to select for accessibility", () => {
```
Comment (bangbay-bluetiger) [inline services/ui-src/src/components/fields/DropdownField.test.tsx:323]: Nit: I would move this block before `testA11yAct`, as that's usually the last block in tests.

File: services/ui-src/src/components/fields/DropdownField.test.tsx [hunk 5]
```diff
@@ -313,4 +313,34 @@ describe("<DropdownField />", () => {
...
+          errors: {
+            testDropdown: { value: { message: "A response is required" } },
+          },
+        },
+      }));
+
+      render(dropdownComponentWithOptions);
+      const dropdown = screen.getByLabelText("test-dropdown-label");
+
+      expect(dropdown).toHaveAttribute("aria-invalid", "true");
```
Comment (bangbay-bluetiger) [inline services/ui-src/src/components/fields/DropdownField.test.tsx:338]: I would add a test for the negative case. You could probably sneak it into one of the other test blocks:
```
expect(dropdown).toHaveAttribute("aria-invalid", "false");
```

File: services/ui-src/src/components/fields/DropdownField.tsx [hunk 6]
```diff
@@ -214,17 +222,18 @@ export const DropdownField = ({
       <Label htmlFor={name} id={`${name}-label`}>
         {labelText || ""}
       </Label>
-      {parsedHint && <Hint id={name}>{parsedHint}</Hint>}
-      {errorMessage && <InlineError>{errorMessage}</InlineError>}
+      {parsedHint && <Hint id={`${name}-hint`}>{parsedHint}</Hint>}
```
Comment (bangbay-bluetiger) [inline services/ui-src/src/components/fields/DropdownField.tsx:225]: I would also add a test for the hint. You can add a hint prop here:

https://github.com/Enterprise-CMCS/macpro-mdct-mcr/blob/b2ed1d403c0fbdee076b555c94fe3b43bfef4345/services/ui-src/src/components/fields/DropdownField.test.tsx#L46-L52

Then your assertion can be something like:
```
expect(dropdown).toHaveAttribute("aria-describedby", "testDropdown__error testDropdown-hint");
```

File: services/ui-src/src/components/fields/DropdownField.tsx [hunk 7]
```diff
@@ -204,7 +210,9 @@ export const DropdownField = ({
   const formErrorState: AnyObject = form?.formState?.errors;
   const errorMessage = formErrorState?.[name]?.value.message;
   const parsedHint = hint && parseCustomHtml(hint);
+  const ariaDescribedBy = `${name}__error${parsedHint ? ` ${name}-hint` : ""}`;
   const nestedChildClasses = nested ? "nested ds-c-choice__checkedChild" : "";
+  const selectClasses = `ds-c-field${errorMessage ? " ds-c-field--error" : ""}`;
```
Comment (bangbay-bluetiger) [inline services/ui-src/src/components/fields/DropdownField.tsx:215]: I'd test for this, too. Could add this to your test block:
```
expect(dropdown).toHaveClass("ds-c-field ds-c-field--error");
```

**Review by bangbay-bluetiger — COMMENTED** [review]
Thanks for adding the test case! Some suggestions to make it more robust.

**Comment by amanda-forrester** [comment]
> Thanks for adding the test case! Some suggestions to make it more robust.

@bangbay-bluetiger Awesome, thank you! I'm still adjusting to writing FE tests, so I really appreciate the tips!

File: services/ui-src/src/components/fields/DropdownField.tsx (hunk 6 above)
Comment (amanda-forrester) [reply services/ui-src/src/components/fields/DropdownField.tsx:225]: @bangbay-bluetiger I don't think it will break anything by adding the hint prop to `dropdownComponentWithOptions` but it feels a bit clearer to me to go ahead and make a new `const dropdownComponentWithHint` and then add a separate test for the hint text, just to keep the hint testing isolated. Thoughts?

Comment (bangbay-bluetiger) [reply services/ui-src/src/components/fields/DropdownField.tsx:225]: That sounds fine to me. You could also an assertion for when there's no hint.

**Comment by amanda-forrester** [comment]
@bangbay-bluetiger I updated the tests!

**Review by bangbay-bluetiger — APPROVED** [review]
🎉

**Review by cy-by — APPROVED** [review]
Error message correctly being read via `aria-describedby`, tested in [deployed env](https://d3ftx38x62gu64.cloudfront.net/).

Image

**Review by ntsummers1 — APPROVED** [review]
